dash/testing/browser.py

```diff
@@ -0,0 +1,281 @@
+        )
+
+    def wait_for_page(self, url=None, timeout=10):
+
+        self.driver.get(self.server_url if url is None else url)
+        try:
+            self.wait_for_element_by_css_selector(
+                self.dash_entry_locator, timeout=timeout
+            )
+        except TimeoutException:
+            logger.exception(
+                "dash server is not loaded within %s seconds", timeout
+            )
+            logger.debug(self.get_logs())
+            raise DashAppLoadingError(
+                "the expected Dash react entry point cannot be loaded"
+                " in browser\n HTML => {}\n Console Logs => {}\n".format(
+                    self.driver.find_element_by_tag_name("body").get_property(
+                        "innerHTML"
+                    ),
+                    "\n".join((str(log) for log in self.get_logs())),
+                )
+            )
+
+    def get_webdriver(self, remote):
+        return (
+            getattr(self, "_get_{}".format(self._browser))()
+            if remote is None
+            else webdriver.Remote(
+                command_executor=remote,
+                desired_capabilities=getattr(
+                    DesiredCapabilities, self._browser.upper()
+                ),
+            )
+        )
+
+    @staticmethod
+    def _get_chrome():
+        options = Options()
+        options.add_argument("--no-sandbox")
+
+        capabilities = DesiredCapabilities.CHROME
+        capabilities["loggingPrefs"] = {"browser": "SEVERE"}
+
+        if "DASH_TEST_CHROMEPATH" in os.environ:
+            options.binary_location = os.environ["DASH_TEST_CHROMEPATH"]
+
+        chrome = webdriver.Chrome(
+            options=options, desired_capabilities=capabilities
+        )
+        chrome.set_window_position(0, 0)
+        return chrome
+
+    @staticmethod
+    def _get_firefox():
+
+        capabilities = DesiredCapabilities.FIREFOX
+        capabilities["loggingPrefs"] = {"browser": "SEVERE"}
+        capabilities["marionette"] = True
+
+        # https://developer.mozilla.org/en-US/docs/Download_Manager_preferences
+        fp = webdriver.FirefoxProfile()
+
+        # this will be useful if we wanna test download csv or other data
+        # files with selenium
+        # TODO this can be fed as a tmpdir fixture from pytest
+        fp.set_preference("browser.download.dir", "/tmp")
+        fp.set_preference("browser.download.folderList", 2)
+        fp.set_preference("browser.download.manager.showWhenStarting", False)
+
+        return webdriver.Firefox(fp, capabilities=capabilities)
+
+    @staticmethod
+    def _is_windows():
+        return sys.platform == "win32"
+
+    def multiple_click(self, css_selector, clicks):
+        for _ in range(clicks):
+            self.find_element(css_selector).click()
+
+    def js_click(self, elem):
+        """click in native javascript way
+        note: this is NOT the recommended way to click"""
```

Comment:
I don't see this used anywhere, is it? Is there a known reason you'd want to use this instead of `find_element().click()`?

Reply:
Yes this is a tricky one. Might not be useful at all for dash. I will remove it.